File: plugins/htprotect/android/src/main/java/com/netease/htp/HtprotectPlugin.java
package com.netease.htp;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Looper;
import android.os.Handler;
import android.text.TextUtils;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import com.netease.htprotect.*;
import com.netease.htprotect.result.AntiCheatResult;

import java.util.HashMap;
import java.util.Map;

/**
 * HtprotectPlugin
 */
public class HtprotectPlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    private Context context;

    private static final String CODE_KEY = "code";
    private static final String MSG_KEY = "msg";
    private static final String TOKEN_KEY = "token";


    public static HtprotectPlugin sInstance;

    public HtprotectPlugin() {
        sInstance = this;
    }

    public void callFlutterMethod(String method, Map<String, String> arguments) {
        if (TextUtils.isEmpty(method)) {
            return;
        }
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(() -> channel.invokeMethod(method, arguments));
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "htprotect");
        channel.setMethodCallHandler(this);
        context = flutterPluginBinding.getApplicationContext();
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("init")) {
            init(call, result);
        } else if (call.method.equals("setRoleInfo")) {
            setRoleInfo(call, result);
        } else if (call.method.equals("logOut")) {
            logOut();
        } else if (call.method.equals("getToken")) {
            getToken(call, result);
        } else if (call.method.equals("ioctl")) {
            ioctl(call, result);
        } else if (call.method.equals("getDataSign")) {
            getDataSign(call, result);
        } else if (call.method.equals("safeCommToServer")) {
            safeCommToServer(call, result);
        } else if (call.method.equals("safeCommFromServer")) {
            safeCommFromServer(call, result);
        } else if (call.method.equals("registerTouchEvent")) {
            registerTouchEvent(call);
        } else if (call.method.equals("unregisterTouchEvent")) {
            unregisterTouchEvent();
        } else {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private void init(MethodCall call, Result result) {
        String productId = call.argument("productId");
        com.netease.htprotect.callback.HTPCallback callback = new com.netease.htprotect.callback.HTPCallback() {
            @Override
            public void onReceive(int paramInt, String paramString) {
                HashMap<String, String> map = new HashMap<>();
                map.put(CODE_KEY, String.valueOf(paramInt));
                map.put(MSG_KEY, paramString);
                try {
                    result.success(map);
                } catch (Exception e) {

                }
            }
        };
        HTProtectConfig htProtectConfig = new HTProtectConfig();
        HashMap config = call.argument("params");
        if (!config.isEmpty()) {
            if (config.containsKey("serverType")) {
                Integer intObject = (Integer) config.get("serverType");
                int serverType = intObject.intValue();
                htProtectConfig.setServerType(serverType);
            }
            if (config.containsKey("channel")) {
                String strChannel = (String) config.get("channel");
                htProtectConfig.setChannel(strChannel);
            }
            if (config.containsKey("gameKey")) {
                String strGameKey = (String) config.get("gameKey");
                htProtectConfig.setGameKey(strGameKey);
            }
            if (config.containsKey("extData")) {
                HashMap<String, String> strExtData = (HashMap<String, String>) config.get("extData");
                for (Map.Entry<String, String> entry : strExtData.entrySet()) {
                    htProtectConfig.setExtraData(entry.getKey(), entry.getValue());
                }
            }
        }
        HTProtect.init(context, productId, callback, htProtectConfig);
    }

    private void setRoleInfo(MethodCall call, Result result) {
        String businessId = call.argument("businessId");
        String roleId = call.argument("roleId");
        String roleName = call.argument("roleName");
        String roleAccount = call.argument("roleAccount");
        String roleServer = call.argument("roleServer");
        int serverId = call.argument("serverId");
        String gameJson = call.argument("gameJson");
        int ret = HTProtect.setRoleInfo(businessId, roleId, roleName, roleAccount, roleServer, serverId, gameJson);
        try {
            result.success(ret);
        } catch (Exception e) {

        }
    }

    private void logOut() {
        HTProtect.logOut();
    }

    private void getToken(MethodCall call, Result result) {
        int timeout = call.argument("timeout");
        String businessId = call.argument("businessId");
        com.netease.htprotect.callback.GetTokenCallback callback = new com.netease.htprotect.callback.GetTokenCallback() {
            @Override
            public void onResult(AntiCheatResult acResult) {
                HashMap<String, String> map = new HashMap<>();
                map.put(CODE_KEY, String.valueOf(acResult.code));
                map.put(TOKEN_KEY, acResult.token);
                try {
                    result.success(map);
                } catch (Exception e) {

                }
            }
        };
        HTProtect.getTokenAsync(timeout, businessId, callback);
    }

    private void ioctl(MethodCall call, Result result) {
        int request = call.argument("request");
        String data = call.argument("data");
        String ret = HTProtect.ioctl(request, data);
        try {
            result.success(ret);
        } catch (Exception e) {

        }
    }

    private void getDataSign(MethodCall call, Result result) {
        int algIndex = call.argument("algIndex");
        String inputData = call.argument("inputData");
        String ret = HTProtect.getDataSign(inputData, algIndex);
        try {
            result.success(ret);
        } catch (Exception e) {

        }
    }

    private void safeCommToServer(MethodCall call, Result result) {
        int version = call.argument("version");
        int alg = call.argument("alg");
        String input = call.argument("input");
        boolean isCrucial = call.argument("isCrucial");
        com.netease.htprotect.result.SafeCommResult safeCommResult = HTProtect.safeCommToServerV30(version, alg, input.getBytes(), isCrucial);
        try {
            HashMap<String, String> map = new HashMap<>();
            map.put(CODE_KEY, String.valueOf(safeCommResult.ret));
            map.put(MSG_KEY, safeCommResult.encResult);
            result.success(map);
        } catch (Exception e) {

        }
    }

    private void safeCommFromServer(MethodCall call, Result result) {
        int alg = call.argument("alg");
        int timeout = call.argument("timeout");
        String input = call.argument("input");

        com.netease.htprotect.result.SafeCommResult safeCommResult = HTProtect.safeCommFromServer(alg, timeout, input);
        try {
            HashMap<String, String> map = new HashMap<>();
            map.put(CODE_KEY, String.valueOf(safeCommResult.ret));
            String msg = new String(safeCommResult.decResult, "UTF-8");
            map.put(MSG_KEY, msg);
            result.success(map);
        } catch (Exception e) {

        }
    }

    private void registerTouchEvent(MethodCall call) {
        int gameplayId = call.argument("gameplayId");
        int sceneId = call.argument("sceneId");
        HTProtect.registerTouchEvent(gameplayId, sceneId);
    }

    private void unregisterTouchEvent() {
        HTProtect.unregisterTouchEvent();
    }
}

File: plugins/htprotect/ios/Classes/HtprotectPlugin.m
#import "HtprotectPlugin.h"
#import <RiskPerception/NTESRiskUniPerception.h>
#import <RiskPerception/NTESRiskUniConfiguration.h>
#import <RiskPerception/NTESRiskUniUtil.h>


@implementation HtprotectPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"htprotect"
            binaryMessenger:[registrar messenger]];
  HtprotectPlugin* instance = [[HtprotectPlugin alloc] init];
  instance.channel = channel;
  [registrar addApplicationDelegate:instance];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
   
  } else if ([@"init" isEqualToString:call.method]) {
      [self init:call result:result];
  } else if ([@"setRoleInfo" isEqualToString:call.method]) {
      [self setRoleInfo:call result:result];
  } else if ([@"logOut" isEqualToString:call.method]) {
      [self logOut:call result:result];
  } else if ([@"getToken" isEqualToString:call.method]) {
      [self getToken:call result:result];
  } else if ([@"ioctl" isEqualToString:call.method]) {
      [self ioctl:call result:result];
  } else if ([@"registInfoReceiver" isEqualToString:call.method]) {
      [self registInfoReceiver:call result:result];
  } else if ([@"getDataSign" isEqualToString:call.method]) {
      [self getDataSign:call result:result];
  } else if ([@"safeCommToServer" isEqualToString:call.method]) {
      [self safeCommToServer:call result:result];
  } else if ([@"safeCommFromServer" isEqualToString:call.method]) {
      [self safeCommFromServer:call result:result];
  } else {
      result(FlutterMethodNotImplemented);
  }
}

- (void)init:(FlutterMethodCall*) call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    NSString *productId = [arguments objectForKey:@"productId"];
    NSDictionary *params = [arguments objectForKey:@"params"];
    NSString *channel = [arguments objectForKey:@"channel"];
    NSString *gameKey = [arguments objectForKey:@"gameKey"];
    NSString *serverType = [arguments objectForKey:@"serverType"];
    NSDictionary *extDict = [params objectForKey:@"extData"];
    [NTESRiskUniConfiguration setChannel:channel];
    [NTESRiskUniConfiguration setServerType:[serverType integerValue]];
  
    [extDict enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        [NTESRiskUniConfiguration setExtraData:obj forValue:key];
    }];
    
    [[NTESRiskUniPerception fomentBevelDeadengo] init:productId callback:^(int code, NSString * _Nonnull msg, NSString * _Nonnull content) {
        NSString *codeString = [NSString stringWithFormat:@"%d",code];
        resultDict(@{@"code":codeString ?:@""});
    }];
   
}

/// 设置角色信息
- (void)setRoleInfo:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    NSString *businessId = [arguments objectForKey:@"businessId"];
    NSString *roleId = [arguments objectForKey:@"roleId"];
    NSString *roleName = [arguments objectForKey:@"roleName"];
    
    NSString *roleAccount = [arguments objectForKey:@"roleAccount"];
    NSString *roleServer = [arguments objectForKey:@"roleServer"];
    NSString *serverId = [arguments objectForKey:@"serverId"];
    
    NSString *gameJson = [arguments objectForKey:@"gameJson"];

    [[NTESRiskUniPerception fomentBevelDeadengo] setRoleInfo:businessId roleId:roleId?:@"" roleName:roleName?:@"" roleAccount:roleAccount?:@"" roleServer:roleServer?:@"" serverId:[serverId intValue] gameJson:gameJson];
}

- (void)logOut:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    [[NTESRiskUniPerception fomentBevelDeadengo] logOut];
}

- (void)getToken:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    int timeout = [[arguments objectForKey:@"timeout"] intValue];
    NSString *businessId = [arguments objectForKey:@"businessId"];
    [[NTESRiskUniPerception fomentBevelDeadengo] getTokenAsync:businessId?:@"" withTimeout:timeout completeHandler:^(AntiCheatResult * _Nonnull result) {
        resultDict(@{@"code":@(result.code),@"token":result.token?:@""});
    }];
}

- (void)ioctl:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    int request = [[arguments objectForKey:@"request"] intValue];
    NSString *data = [arguments objectForKey:@"data"];
    NSString *result = [[NTESRiskUniPerception fomentBevelDeadengo] ioctl:request withData:data];
    resultDict(result?:@"");
}

- (void)registInfoReceiver:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    [[NTESRiskUniPerception fomentBevelDeadengo] registInfoReceiver:^(int type, NSString * _Nonnull info) {
        NSDictionary *dict = @{@"type":@(type),@"info":info?:@""};
        [self.channel invokeMethod:@"onCallBack" arguments:dict];
    }];
}

- (void)getDataSign:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    NSString *data = [arguments objectForKey:@"inputData"];
    NSString *dataSignString = [NTESRiskUniUtil getDataSign:data];
    resultDict(dataSignString?:@"");
}

- (void)safeCommToServer:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    int alg = [[arguments objectForKey:@"alg"] intValue];
    NSString *input = [arguments objectForKey:@"input"];
    NSData *data = [input dataUsingEncoding:NSUTF8StringEncoding];
    NSError *error;
    NSString *dataSignString = [NTESRiskUniUtil safeCommToServer:alg withData:data error:&error];
    if (!error) {
        resultDict(@{@"msg":dataSignString?:@"",@"code":@(200)});
    } else {
        resultDict(@{@"msg":[error.userInfo objectForKey:@"NSLocalizedDescription"]?:@"",@"code":@(error.code)});
    }
   
}

- (void)safeCommFromServer:(FlutterMethodCall*)call result:(FlutterResult)resultDict {
    NSDictionary *arguments = call.arguments;
    int alg = [[arguments objectForKey:@"alg"] intValue];
    int timeout = [[arguments objectForKey:@"timeout"] intValue];
    NSString *input = [arguments objectForKey:@"input"];
    NSData *data = [input dataUsingEncoding:NSUTF8StringEncoding];
    NSError *error;
    NSData *dataSignData = [NTESRiskUniUtil safeCommFromServer:alg timeout:timeout withData:data error:&error];
    if (!error) {
        NSString *dataSignString = [[NSString alloc] initWithData:dataSignData encoding:NSUTF8StringEncoding];
        resultDict(@{@"msg":dataSignString?:@"",@"code":@(200)});
    } else {
        resultDict(@{@"msg":[error.userInfo objectForKey:@"NSLocalizedDescription"]?:@"",@"code":@(error.code)});
    }
   
}

@end
